### PR TITLE
Adds compile option to configuration

### PR DESCRIPTION
**Pull request by rossta**
This change introduces a new configuration option for enabling lazy compilation via `config/webpacker.yml`. 

Along with #502, this would allow for running a webpack-dev-server process in the test environment instead of enforcing compilation to the `public` directory. The fallback for a missing `compile` key keeps the default lazy-compilation behavior intact.

The `compile` key name borrows from the `assets.compile` option for the asset pipeline.

### CODE REVIEW

**Comment by javan** [comment]
+1 to making this configurable

> this would allow for running a webpack-dev-server process in the test environment instead of enforcing compilation to the public directory

Wouldn't running `webpack-dev-server` prevent lazy compiling on its own since it updates manifest.json? `compile_and_find!(name)` is only called when there's no entry in the manifest.

File: lib/install/config/webpacker.yml [hunk 1]
```diff
@@ -4,6 +4,7 @@ default: &default
   source_path: app/javascript
   source_entry_path: packs
   public_output_path: packs
+  compile: false
```
Comment (javan) [inline lib/install/config/webpacker.yml]: WDYT about flipping this around? Make it `true` by default and `false` in production.

File: lib/webpacker/configuration.rb [hunk 2]
```diff
@@ -36,6 +36,10 @@ def source
       fetch(:source_path)
     end
 
+    def compile?
+      fetch(:compile).nil? ? Webpacker.env.test? : fetch(:compile)
```
Comment (javan) [inline lib/webpacker/configuration.rb]: I'd drop the `Webpacker.env.test?` default condition and just use the configuration option.

File: lib/install/config/webpacker.yml (hunk 1 above)
Comment (javan) [inline lib/install/config/webpacker.yml]: Since compilation only happens when packs are missing (vs. stale), the name `compile` doesn't feel quite right. Perhaps something more descriptive like `compile_missing_packs`?

**Comment by rossta** [comment]
@javan Thanks for the feedback.

> Wouldn't running webpack-dev-server prevent lazy compiling on its own since it updates manifest.json? 

You're right—it would've been more accurate for me to say the "lookup and compile if not found" behavior is only possible in `Webpacker.env.test` prior to these changes.

File: lib/install/config/webpacker.yml (hunk 1 above)
Comment (rossta) [inline lib/install/config/webpacker.yml]: 👍 Flipping the values and making the name more descriptive both sound like good changes.

File: lib/webpacker/configuration.rb (hunk 2 above)
Comment (rossta) [inline lib/webpacker/configuration.rb]: One concern is for folks upgrading for whom the `compile_missing_packs` entry would be missing... in this case, they would have to add in the values to get the expected behavior back in `test`. Should we add a warning or upgrade note in README?

Comment (gauravtiwari) [inline lib/webpacker/configuration.rb]: @rossta Yes please 👍  add it to Chagelog and highlight if it's a breaking change

File: CHANGELOG.md [hunk 3]
```diff
@@ -1,5 +1,20 @@
-## [2.0] - 2017-05-24
+## Unreleased
+
+### Fixed
+- Update `webpack-dev-server.tt` to respect RAILS_ENV and NODE_ENV values [#502](https://github.com/rails/webpacker/issues/502)
```
Comment (javan) [inline CHANGELOG.md]: Is this entry related to a change in this PR?

Comment (rossta) [inline CHANGELOG.md]: Loosely. It was a necessary change along with this PR for the ability to run the webpack-dev-server in test. Would you prefer a separate PR for this entry?

**Comment by javan** [comment]
Thank you!
